xdsclient: fix new watcher hang when registering for removed resource

purnesh42H opened the pull request:
Addresses: #7819

Right now if any new watcher try to register watcher for removed resource, it hangs as there is no update to be received. This fix adds a condition in xDS Client to trigger watch callback `OnResourceDoesNotExist()` to new watcher if its registering for a removed resource.

RELEASE NOTES:

- xds: fixed an edge-case issue where some clients or servers would not initialize correctly if another channel or server with the same target was already in use.

easwars commented:
Same here. I think this is also worth release noting.

easwars commented on xds/internal/xdsclient/tests/lds_watchers_test.go:
```diff
@@ -742,7 +742,9 @@ func (s) TestLDSWatch_ValidResponseCancelsExpiryTimerBehavior(t *testing.T) {
 //  1. Removing a resource should trigger the watch callback with a resource
 //     removed error. It should not trigger the watch callback for an unrelated
-//     resource.
+//     resource. If any new watcher try to register watcher for removed
```
I know that we haven't always done a good job with keeping tests focussed on one thing. So some of our tests do multiple things. But it would be nice to have a separate test for this new scenario instead of adding more to an existing test (which looking at it now, is already doing a lot).

See: https://engdoc.corp.google.com/eng/doc/tott/episodes/726.md?cl=head

purnesh42H replied:
Added a separate test

easwars commented on xds/internal/xdsclient/tests/lds_watchers_test.go:
```diff
@@ -871,6 +871,90 @@ func (s) TestLDSWatch_ResourceRemoved(t *testing.T) {
+	}
+
+	// The existing watcher should receive a resource removed error.
+	if err := verifyListenerUpdate(ctx, lw1.updateCh, listenerUpdateErrTuple{
+		err: xdsresource.NewErrorf(xdsresource.ErrorTypeResourceNotFound, ""),
+	}); err != nil {
+		t.Fatal(err)
+	}
```
This is a little hard to read. Could we do something like

```golang
	updateErr := listenerUpdateErrTuple{err: xdsresource.NewErrorf(xdsresource.ErrorTypeResourceNotFound, "")}
	if err := verifyListenerUpdate(ctx, lw1.updateCh, updateErr); err != nil {
		t.Fatal(err)
	}
```

easwars commented on xds/internal/xdsclient/tests/lds_watchers_test.go:
```diff
@@ -871,6 +871,90 @@ func (s) TestLDSWatch_ResourceRemoved(t *testing.T) {
+		t.Fatal(err)
+	}
+
+	// Any new watcher trying to register for removed resource, should get
+	// resource removed error as well.
```
Nit: Maybe something like

New watchers attempting to register for a deleted resource should also receive a "resource removed" error.

purnesh42H replied:
yes, better. Done

easwars commented on xds/internal/xdsclient/tests/lds_watchers_test.go:
```diff
@@ -871,6 +871,90 @@ func (s) TestLDSWatch_ResourceRemoved(t *testing.T) {
+	lw2 := newListenerWatcher()
+	ldsCancel2 := xdsresource.WatchListener(client, resourceName, lw2)
+	defer ldsCancel2()
+	if err := verifyListenerUpdate(ctx, lw2.updateCh, listenerUpdateErrTuple{
+		err: xdsresource.NewErrorf(xdsresource.ErrorTypeResourceNotFound, ""),
+	}); err != nil {
+		t.Fatal(err)
+	}
```
Same here as the earlier comment.

purnesh42H replied:
Updated

purnesh42H commented on xds/internal/xdsclient/tests/lds_watchers_test.go:
```diff
@@ -871,6 +871,90 @@ func (s) TestLDSWatch_ResourceRemoved(t *testing.T) {
 }
 
+// TestLDSWatch_NewWatcherForRemovedResource covers the case where a new
+// watcher registers for the resource which is removed. The test verifies
```
Done

purnesh42H commented on xds/internal/xdsclient/tests/lds_watchers_test.go:
```diff
@@ -871,6 +871,90 @@ func (s) TestLDSWatch_ResourceRemoved(t *testing.T) {
+// TestLDSWatch_NewWatcherForRemovedResource covers the case where a new
+// watcher registers for the resource which is removed. The test verifies
+// the following scenarios:
+//  1. Removing a resource should trigger the watch callback with a resource
```
yeah looks better. Thanks

purnesh42H commented on xds/internal/xdsclient/tests/lds_watchers_test.go:
```diff
@@ -871,6 +871,90 @@ func (s) TestLDSWatch_ResourceRemoved(t *testing.T) {
+	defer close()
+
+	// Register watch for the listener resources and have the
+	// callbacks push the received updates on to a channel.
```
yeah had changed everywhere. May be missed this. Done

purnesh42H commented on xds/internal/xdsclient/tests/lds_watchers_test.go:
```diff
@@ -871,6 +871,90 @@ func (s) TestLDSWatch_ResourceRemoved(t *testing.T) {
+
+	// Register watch for the listener resources and have the
+	// callbacks push the received updates on to a channel.
+	resourceName := ldsName
```
not really. Had kept for readability. Removed

easwars commented:
For the release note, how about something like:

- xds: fixed an issue in the xDS client where new watchers would hang when registering a watch for a deleted resource
- xdsclient: fixed  an issue where new watchers would hang when registering a watch for a deleted resource

dfawley commented:
> For the release note, how about something like:

Nobody is using our xdsclient so we don't want to document it this way.  Could this affect grpc users?  If so, how?  Something like:

* xds: fixed an issue where some clients or servers would not initialize correctly if another channel or server with the same target was already in use

?

purnesh42H commented:
> Nobody is using our xdsclient so we don't want to document it this way. Could this affect grpc users? If so, how? Something like:

yeah initially i didn't add release notes because i thought its not being directly used by users. I guess its just used as part of xds lb policy so may be having xds as top level package is fine.

> fixed an issue where some clients or servers would not initialize correctly if another channel or server with the same target was already in use

this is not clear to me. If someone is using xds lb policy, they would know about resource deletions right? So, its fine to mention that? May be something like, `the client initialization can delay for xds resources if the resource that existed previously got deleted`?

purnesh42H commented:
@dfawley ping on the release notes

> > Nobody is using our xdsclient so we don't want to document it this way. Could this affect grpc users? If so, how? Something like:
> 
> yeah initially i didn't add release notes because i thought its not being directly used by users. I guess its just used as part of xds lb policy so may be having xds as top level package is fine.
> 
> > fixed an issue where some clients or servers would not initialize correctly if another channel or server with the same target was already in use
> 
> this is not clear to me. If someone is using xds lb policy, they would know about resource deletions right? So, its fine to mention that? May be something like, `the client initialization can hang for xds resources if the resource that existed previously got deleted`?

@dfawley

purnesh42H commented:
> > Nobody is using our xdsclient so we don't want to document it this way. Could this affect grpc users? If so, how? Something like:
> 
> yeah initially i didn't add release notes because i thought its not being directly used by users. I guess its just used as part of xds lb policy so may be having xds as top level package is fine.
> 
> > fixed an issue where some clients or servers would not initialize correctly if another channel or server with the same target was already in use
> 
> this is not clear to me. If someone is using xds lb policy, they would know about resource deletions right? So, its fine to mention that? May be something like, `the client initialization can hang for xds resources if the resource that existed previously got deleted`?

@dfawley

easwars commented:
> this is not clear to me. If someone is using xds lb policy, they would know about resource deletions right? So, its fine to mention that? May be something like, the client initialization can hang for xds resources if the resource that existed previously got deleted?

The way users use xDS on the client is by specifying a dial target with an `xds:///` or `xdstp:///` scheme, and by importing the `xds` package for side-effects. The resources are watched by the xDS resolver and the LB policies and these are not directly exposed to the user.

And the way users use xDS on the server is by explicitly created an xDS-enabled gRPC server. This functionality is provided by the `xds` package. Here as well, the actual resources are watched by the server implementation and not directly exposed to the user.

So, the user will not see the effects on watch callbacks being invoked or not, but they would be able to see the knock-on effects of those.

So yes, I agree with @dfawley in making the release note more applicable from the user's pov. But I feel the proposal from Doug made it sounds more scary than it actually is. So maybe?

xds: fixed a (rare|edge-case|corner-case) issue where some clients or servers would not initialize correctly if another channel or server with the same target was already in use

purnesh42H commented:
> > this is not clear to me. If someone is using xds lb policy, they would know about resource deletions right? So, its fine to mention that? May be something like, the client initialization can hang for xds resources if the resource that existed previously got deleted?
> 
> The way users use xDS on the client is by specifying a dial target with an `xds:///` or `xdstp:///` scheme, and by importing the `xds` package for side-effects. The resources are watched by the xDS resolver and the LB policies and these are not directly exposed to the user.
> 
> And the way users use xDS on the server is by explicitly created an xDS-enabled gRPC server. This functionality is provided by the `xds` package. Here as well, the actual resources are watched by the server implementation and not directly exposed to the user.
> 
> So, the user will not see the effects on watch callbacks being invoked or not, but they would be able to see the knock-on effects of those.
> 
> So yes, I agree with @dfawley in making the release note more applicable from the user's pov. But I feel the proposal from Doug made it sounds more scary than it actually is. So maybe?
> 
> xds: fixed a (rare|edge-case|corner-case) issue where some clients or servers would not initialize correctly if another channel or server with the same target was already in use

Updated the release notes as edge case. We can revisit during release again. Merging now

dfawley commented:
I'm not really concerned with the release notes sounding too scary, but they should be accurate.  So it should clearly indicate that it's a "sometimes" not an "always".  But "an edge-case issue" sounds awkward to me.

* xds: fixed a bug causing clients or servers to hang at startup if the same target is already in use and some of its resources are not available.

?
